README.md
# blackjack

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar blackjack-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2023 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-12-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-12-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/blackjack/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/blackjack/compare/0.1.0...0.1.1


test/business_test/business_test.clj
(ns business-test.business-test
  (:require [midje.sweet :refer :all]
            [business.business :refer :all :as b]
            [cheat-sheet.cheat-sheet :refer :all :as sheet]))
(fact "Testing generate-card function"
      (let [card (generate-card "10" "heart")]
        (fact "Generated card should have the correct value and suit"
              (card :value) => "10"
              (card :suit) => "heart")))

(fact "Testing 'get-nth-player-card' function"
      (fact "When card has the nth player card"
            (let [test-card {:card-1 {:value "10" :suit "heart"}
                             :card-2 {:value "A" :suit "diamond"}}
                  result (get-nth-player-card test-card 2)]
              result => {:value "A" :suit "diamond"}))

      (fact "When card has no nth player card"
            (let [test-card {:card-1 {:value "10" :suit "heart"}}
                  result (get-nth-player-card test-card 2)]
              result => falsey))

      (fact "When card is nil"
            (let [result (get-nth-player-card nil 2)]
              result => falsey)))
(fact "Testing 'get-dealer-card' function"
      (fact "When card has a dealer card"
            (let [test-card {:dealer-card [{:value "10" :suit "heart"}]}
                  result (get-dealer-card test-card)]
              result => {:value "10" :suit "heart"}))

      (fact "When card has no dealer card"
            (let [test-card {:dealer-card []}
                  result (get-dealer-card test-card)]
              result => falsey))

      (fact "When card is nil"
            (let [result (get-dealer-card nil)]
              result => falsey)))

(fact "Testing add-new-player-card function edge cases"
      (fact "When adding a new card to an empty hand"
            (let [test-hand {}
                  new-card {:value "10" :suit "heart"}
                  result (add-new-player-card test-hand new-card)]
              result => {:card-1 {:value "10" :suit "heart"}}))

      (fact "When adding multiple cards to a hand"
            (let [test-hand {}
                  new-card-1 {:value "10" :suit "heart"}
                  new-card-2 {:value "A" :suit "diamond"}
                  new-card-3 {:value "5" :suit "club"}
                  result (-> test-hand
                             (add-new-player-card new-card-1)
                             (add-new-player-card new-card-2)
                             (add-new-player-card new-card-3))]
              result => {:card-1 {:value "10" :suit "heart"}
                         :card-2 {:value "A" :suit "diamond"}
                         :card-3 {:value "5" :suit "club"}}))

      (fact "When adding cards with special characters"
            (let [test-hand {}
                  new-card-1 {:value "10" :suit "heart"}
                  new-card-2 {:value "A" :suit "diamond"}
                  new-card-3 {:value "*" :suit "@"}
                  result (-> test-hand
                             (add-new-player-card new-card-1)
                             (add-new-player-card new-card-2)
                             (add-new-player-card new-card-3))]
              result => {:card-1 {:value "10" :suit "heart"}
                         :card-2 {:value "A" :suit "diamond"}
                         :card-3 {:value "*" :suit "@"}})))
(fact "'add-both' should return 'nil' when provided with dereferenced values."
      (let [player-cards ({:card-1 {:value "10" :suit "heart"}}
                          {:card-2 {:value "10" :suit "heart"}})
            current-cards {:value "7" :suit "club"}]
        (b/add-both player-cards current-cards) => falsey))
(fact "Tests edge cases for 'card-value' function."
      (card-value "jack") => 10
      (card-value "unknown") => falsey)

(fact "Updating player card value with a non-existing card number should return nil."
      (let [player-cards ({:card-1 {:value "10" :suit "heart"}}
                          {:card-2 {:value "10" :suit "heart"}})]
        (update-player-card-value player-cards 3 "5") => falsey))

(fact "Updating player card value."
      (let [player-cards (atom {:card-1 {:value "10" :suit "heart"}
                                :card-2 {:value "10" :suit "heart"}})]
        (fact "Updating player card value with a non-existing card number should return nil."
              (b/update-player-card-value! player-cards 3 "5") => falsey)
        (fact "'update-player-card-value!' should return 'nil' when provided with dereferenced values."
              (b/update-player-card-value! @player-cards 3 "5") => falsey)))

(fact "Calculating sum of player cards"
      (let [empty-hand {}
            single-card {:card-1 {:value "10" :suit "heart"}}]

        (fact "With an empty hand should return 0."
              (player-sum empty-hand) => 0)

        (fact "With a single card should return the value of that card."
              (player-sum single-card) => 10)))

(fact "'get-ace-position' and 'get-no-ace-position' should handle different scenarios"
      (let [empty-hand {}
            hand-without-ace {:card-1 {:value "king" :suit "heart"}
                              :card-2 {:value "queen" :suit "diamond"}}
            hand-with-ace {:card-1 {:value "ace" :suit "heart"}
                           :card-2 {:value "king" :suit "diamond"}
                           :card-3 {:value "ace" :suit "spade"}}]
        (fact "Return an empty list for an empty hand."
              (get-ace-position empty-hand) => []
              (get-no-ace-position empty-hand) => [])

        (fact "Return an empty list for a hand with no 'ace' values."
              (get-ace-position hand-without-ace) => []
              (get-no-ace-position hand-without-ace) => [2 1])

        (fact "Return the positions of cards with 'ace' value."
              (get-ace-position hand-with-ace) => [3 1]
              (get-no-ace-position hand-with-ace) => [2])))

(fact "'adjust-ace-value' should handle different scenarios"
      (let [no-ace-hand (atom {:card-1 {:value "king" :suit "heart"}
                               :card-2 {:value "queen" :suit "diamond"}})
            two-ace-hand (atom {:card-1 {:value "ace" :suit "heart"}
                                :card-2 {:value "ace" :suit "diamond"}})
            ace-value-eleven (atom {:card-1 {:value "ace" :suit "heart"}
                                    :card-2 {:value "king" :suit "diamond"}})
            ace-value-one (atom {:card-1 {:value "ace" :suit "heart"}
                                 :card-2 {:value "king" :suit "diamond"}
                                 :card-3 {:value "ace" :suit "spade"}})]

        (fact "No adjustment is made for a hand without 'ace' cards."
              (adjust-ace-value! no-ace-hand) => (deref no-ace-hand))

        (fact "No adjustment is made when the sum is less or equals to 21."
              (adjust-ace-value! no-ace-hand) => (deref no-ace-hand)
              (adjust-ace-value! ace-value-eleven) => (deref ace-value-eleven))

        (fact "Adjust 'ace' value from 11 to 1 when the sum is over 21."
              (adjust-ace-value! ace-value-one) => {:card-1 {:value 1 :suit "heart"}
                                                    :card-2 {:value "king" :suit "diamond"}
                                                    :card-3 {:value 1 :suit "spade"}}
              (adjust-ace-value! two-ace-hand) => {:card-1 {:value "ace" :suit "heart"}
                                                   :card-2 {:value 1 :suit "diamond"}})))


(fact "Testing between when number is outside the range."
      (b/between? 5 10 15) => falsey)

(fact "Testing ace-to-A function"
      (ace-to-A "ace") => "A"
      (ace-to-A "king") => "king"
      (ace-to-A "invalid") => "invalid"
      (ace-to-A nil) => falsey)

(fact "Testing 'recommend-move' function"
      (let [cheat-sheet @sheet/blackjack-cheat-sheet
            current-cards {:player-cards (list {:value "ace" :suit "heart"} {:value "king" :suit "heart"})
                           :dealer-card  (list {:value "king" :suit "heart"})}
            player-hand {:card-1 {:value "ace" :suit "heart"}
                         :card-2 {:value "king" :suit "heart"}}]

        (fact "Player has a Blackjack"
              (b/recommend-move cheat-sheet current-cards player-hand) => "Blackjack!"))

      (let [cheat-sheet @sheet/blackjack-cheat-sheet
            current-cards {:player-cards (list {:value "3" :suit "heart"} {:value "2" :suit "heart"} {:value "3" :suit "heart"})
                           :dealer-card  (list {:value "king" :suit "heart"})}
            player-hand {:card-1 {:value "3" :suit "heart"}
                         :card-2 {:value "2" :suit "heart"}
                         :card-3 {:value "3" :suit "heart"}}]
        (fact "Player has more than 2 cards and current sum is less than 9"
              (b/recommend-move cheat-sheet current-cards player-hand) => "H"))

      (fact "Player has 2 cards, and their sum is between 8 and 18, and both cards are different, and no Ace in the hand"
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {:player-cards (list {:value "5" :suit "heart"} {:value "4" :suit "heart"})
                                 :dealer-card  (list {:value "king" :suit "heart"})}
                  player-hand {:card-1 {:value "5" :suit "heart"}
                               :card-2 {:value "4" :suit "heart"}}]

              (b/recommend-move cheat-sheet current-cards player-hand) => "H"))

      (fact "Player has 2 Aces."
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {:player-cards (list {:value "ace" :suit "heart"} {:value "ace" :suit "heart"})
                                 :dealer-card  (list {:value "king" :suit "heart"})}
                  player-hand {:card-1 {:value "ace" :suit "heart"}
                               :card-2 {:value "ace" :suit "heart"}}]

              (b/recommend-move cheat-sheet current-cards player-hand) => "P"))

      (fact "Player has 1 Ace."
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {:player-cards (list {:value "ace" :suit "heart"} {:value "4" :suit "heart"})
                                 :dealer-card  (list {:value "king" :suit "heart"})}
                  player-hand {:card-1 {:value "ace" :suit "heart"}
                               :card-2 {:value "4" :suit "heart"}}]

              (b/recommend-move cheat-sheet current-cards player-hand) => "H"))

      (fact "Player has 2 cards of the same value (not Ace)."
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {:player-cards (list {:value "4" :suit "heart"} {:value "4" :suit "heart"})
                                 :dealer-card  (list {:value "king" :suit "heart"})}
                  player-hand {:card-1 {:value "4" :suit "heart"}
                               :card-2 {:value "4" :suit "heart"}}]

              (b/recommend-move cheat-sheet current-cards player-hand) => "H"))

      (fact "Player has 2 cards, and current sum is greater than 17."
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {:player-cards (list {:value "10" :suit "heart"} {:value "9" :suit "heart"})
                                 :dealer-card  (list {:value "king" :suit "heart"})}
                  player-hand {:card-1 {:value "10" :suit "heart"}
                               :card-2 {:value "9" :suit "heart"}}]

              (b/recommend-move cheat-sheet current-cards player-hand) => "S"))

      (fact "Empty hand. (Not covered in the cheat sheet)"
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {}
                  player-hand {}]
              (b/recommend-move cheat-sheet current-cards player-hand) => falsey))

      (fact "Player has 3 cards, and current sum is greater than 21. (Not covered in the cheat sheet)"
            (let [cheat-sheet @sheet/blackjack-cheat-sheet
                  current-cards {:player-cards (list {:value "10" :suit "heart"} {:value "9" :suit "heart"} {:value "5" :suit "heart"})
                                 :dealer-card  (list {:value "king" :suit "heart"})}
                  player-hand {:card-1 {:value "10" :suit "heart"}
                               :card-2 {:value "9" :suit "heart"}
                               :card-3 {:value "5" :suit "heart"}}]
              (b/recommend-move cheat-sheet current-cards player-hand) => falsey)))













test/cheat_sheet_test/cheat_sheet_test.clj
(ns cheat-sheet-test.cheat-sheet-test
  (:require [midje.sweet :refer :all]
            [cheat-sheet.cheat-sheet :refer :all :as cs]))

(fact "Tests cheat sheet values conversion."
      (cs/value-for-ch "ace") => "A"
      (cs/value-for-ch "jack") => "10"
      (cs/value-for-ch "2") => "2"
      (cs/value-for-ch "unknown") => "unknown"
      (cs/value-for-ch "Ace") => "A"
      (cs/value-for-ch " ace ") => "A"
      (cs/value-for-ch 2) => 2
      (cs/value-for-ch "") => "")


(fact "'update-sheet' should update value at a given position."
      (let [initial-cheat-sheet (atom
                                  {:players-cards ["8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "A2" "A3" "A4" "A5" "A6" "A7" "A8" "A9"
                                                   "22" "33" "44" "55" "66" "77" "88" "99" "1010" "AA"]
                                   :dealers-card  ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"]
                                   :values        (vec (repeat 28 (vec (repeat 10 "H"))))})]
        (fact "Updates value at a valid position"
              (get-in (:values (cs/update-sheet @initial-cheat-sheet 2 3 "new-value")) [2 3]) => "new-value")
        (fact "Does not update for non-existing indexes"
              (cs/update-sheet @initial-cheat-sheet 100 100 "new-value") => falsey)))

(fact "Tests exception handling in case of non-existing player's and dealer's card."
      (let
        [initial-cheat-sheet (atom
                               {:players-cards ["8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "A2" "A3" "A4" "A5" "A6" "A7" "A8" "A9"
                                                "22" "33" "44" "55" "66" "77" "88" "99" "1010" "AA"]
                                :dealers-card  ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"]
                                :values        (vec (repeat 28 (vec (repeat 10 "H"))))})]
        (cs/get-move @initial-cheat-sheet "invalid-row" "invalid-col") => falsey))

(fact "Checks setting cheat sheet values."
      (let [initial-cheat-sheet (atom
                                  {:players-cards ["8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "A2" "A3" "A4" "A5" "A6" "A7" "A8" "A9"
                                                   "22" "33" "44" "55" "66" "77" "88" "99" "1010" "AA"]
                                   :dealers-card  ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"]
                                   :values        (vec (repeat 28 (vec (repeat 10 "H"))))})]

        (fact "Sets single column value"
              (cs/set-value! "8" "2" "new-value" initial-cheat-sheet)
              (cs/get-move @initial-cheat-sheet "8" "2") => "new-value")

        (fact "Sets multiple columns values"
              (cs/set-value! "8" ["2" "3" "4"] "new-value" initial-cheat-sheet)
              (map #(cs/get-move @initial-cheat-sheet "8" %) ["2" "3" "4"]) => ["new-value" "new-value" "new-value"])

        (fact "Does not set value in case of non-existing row and column"
              (cs/set-value! "invalid-row" "invalid-col" "new-value" initial-cheat-sheet)
              (cs/get-move @initial-cheat-sheet "invalid-row" "invalid-col") => falsey)))


(fact "Tests update-cheat-sheet-cell."
      (let [initial-cheat-sheet (atom
                                  {:players-cards ["8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "A2" "A3" "A4" "A5" "A6" "A7" "A8" "A9"
                                                   "22" "33" "44" "55" "66" "77" "88" "99" "1010" "AA"]
                                   :dealers-card  ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"]
                                   :values        (vec (repeat 28 (vec (repeat 10 "H"))))})
            all-rows (cs/get-rows @initial-cheat-sheet)
            all-cols (cs/get-cols @initial-cheat-sheet)]
        (cs/update-cheat-sheet-cell all-rows all-cols initial-cheat-sheet "9" ["3" "4" "5" "6"] "DD")
        (cs/get-move @initial-cheat-sheet "9" "4") => "DD"
        (cs/update-cheat-sheet-cell all-rows all-cols initial-cheat-sheet "invalid-row" ["3" "4" "5" "6"] "DD") => falsey
        (cs/update-cheat-sheet-cell all-rows all-cols initial-cheat-sheet "9" "invalid-col" "DD") => falsey))



src/probability/probability.clj
(ns probability.probability)

(defn contain-element?
  "Checks if a specified element is present in the given list."
  [l e]
  (some #(= % e) l))

(defn generate-list
  "Generates a list of integers starting from 1 up to the specified number `n` (inclusive)."
  [n]
  (loop [i 1 num n l '()]
    (if (= (+ n 1) i)
      l
      (recur (inc i) num (cons i l)))))

(defn get-all-values
  [hand]
  "Returns list of both dealer and player original cards values."
  (concat (map #(get % :value) (:player-cards hand))
          [(str (:value (first (:dealer-card hand))))]))

(defn decrement-counter-on-match
  "Decrements the values in a counter atom based on the matching elements between
  a generated list and the values obtained from a card. The list of matching elements
  is determined by the specified fit-value."
  [counter card fit-value]
  (let [l2 (get-all-values card)
        l1 (generate-list fit-value)]
    (swap! counter
           (fn
             [c]
             (reduce (fn
                       [acc elem]
                       (if (contain-element? l1 elem)
                         (dec acc)
                         acc))
                     c
                     l2)))))
(defn count-probability-hit
  "Scenario 1: Recommended move: 'H'
  Calculates the probability of player not busting based on the count of fit-values (counter) and
  the total count (divisor). Returns the result as a formatted percentage."
  [counter divisor card fit-value]
  (str (format "%.2f" (* 100 (float (/ (decrement-counter-on-match counter card fit-value)
                                       divisor)))) "% of not busting."))

(defn value
  "Converts a numeric card value to its corresponding string representation."
  [val]
  (case val
    11 "ace"
    12 "jack"
    13 "queen"
    14 "king"
    val))

(defn subvector
  "Generates a subvector from an input list based on the specified start and stop values."
  [input-list a b]
  (let [start-element (value a)
        stop-element (value b)
        start-index (.indexOf input-list start-element)
        stop-index (inc (.indexOf input-list stop-element))]
    (subvec (vec input-list) start-index stop-index)))

(defn count-probability-s
  "Scenario 2: Recommended move: 'S'
  Calculates the probability of the dealer winning when the player stands."
  [counter divisor current-cards fit-value]
  (float (/ (decrement-counter-on-match counter current-cards fit-value)
            divisor)))

(defn count-probability-stand
  "Calculates the probability of the player winning when the player stands.
  Returns the result as a formatted percentage."
  [counter divisor current-cards fit-value]
  (str (format "%.2f" (* 100 (float (- 1 (count-probability-s counter divisor current-cards fit-value))))) "% of winning."))






src/cheat_sheet/cheat_sheet.clj
(ns cheat-sheet.cheat-sheet)

(def row-num 28)
(def col-num 10)
(def blackjack-cheat-sheet (atom
                             {:players-cards ["8" "9" "10" "11" "12" "13" "14" "15" "16" "17" "A2" "A3" "A4" "A5" "A6" "A7" "A8" "A9"
                                              "22" "33" "44" "55" "66" "77" "88" "99" "1010" "AA"]
                              :dealers-card  ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"]
                              :values        (vec (repeat row-num (vec (repeat col-num "H"))))}))

(defn get-rows
  [cheat-sheet]
  (:players-cards cheat-sheet))
(defn get-cols
  [cheat-sheet]
  (:dealers-card cheat-sheet))
(def row-names (get-rows @blackjack-cheat-sheet))
(def col-names (get-cols @blackjack-cheat-sheet))

(defn value-for-ch
  "Returns a cheat-sheet representation of the card value."
  [val]
  (case (clojure.string/trim (clojure.string/lower-case val))
    "ace" "A"
    "jack" "10"
    "queen" "10"
    "king" "10"
    val))


(defn- get-row-idx
  [row-name cheat-sheet]
  (try
    (.indexOf (:players-cards cheat-sheet) row-name)
       (catch NullPointerException e
         (println "Non-existing player's card.")
         nil)))

(defn- get-col-idx
  [col-name cheat-sheet]
 (try (.indexOf (:dealers-card cheat-sheet) col-name)
      (catch NullPointerException e
        (println "Non-existing dealer's card.")
        nil)))

(defn update-sheet
  "Updates the cheat-sheet matrix by setting the specified value at the given row and column."
  [cheat-sheet row col value]
  (try
    (and
      (<= 0 (dec (count (:values cheat-sheet))))
      (<= 0 (dec (count (first (:values cheat-sheet))))))
    (update-in cheat-sheet [:values row col] (constantly value))
    (catch Exception e
      (println "Non-existing player's or dealer's card.")
      nil)))

(defn get-move
  "Returns move recommendation for player based on player's and dealer's cards."
  [cheat-sheet player-card dealer-card]
  (try
    (get-in (:values cheat-sheet)
            [(get-row-idx player-card cheat-sheet) (get-col-idx (value-for-ch dealer-card) cheat-sheet)])
    (catch NullPointerException e
      (println "Non-existing player's or dealer's card.")
      nil)))

(defn set-value
  "Updates the cheat-sheet matrix by setting the specified value at the given row and multiple columns."
[row col value cheat-sheet]
  (let [row-idx (get-row-idx row cheat-sheet)
        cols (if (vector? col) col [col])]
    (reduce (fn
              [acc col]
              (update-sheet acc row-idx (get-col-idx col cheat-sheet) value))
            cheat-sheet
            cols)))

(defn set-value!
  "Updates the cheat-sheet matrix atom using 'set-value'"
  [row cols value cheat-sheet]
(swap! cheat-sheet (fn
                     [ch]
                     (set-value row cols value ch))))

(defn update-cheat-sheet-cell [all-rows all-cols cheat-sheet row cols value]
  "Updates the cheat-sheet matrix using 'set-value-more-cols!' for desired field located at the intersection
  of the specified row and columns."
  (if (and
        (every? #(some #{%} all-rows) [row])
        (every? #(some #{%} all-cols) cols))
    (set-value! row cols value cheat-sheet)))

(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "9" ["3" "4" "5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "10" ["2" "3" "4" "5" "6" "7" "8" "9"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "11" ["2" "3" "4" "5" "6" "7" "8" "9" "10"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "12" ["4" "5" "6"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "13" ["2" "3" "4" "5" "6"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "14" ["2" "3" "4" "5" "6"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "15" ["2" "3" "4" "5" "6"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "16" ["2" "3" "4" "5" "6"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "17" ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A2" ["5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A3" ["5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A4" ["4" "5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A5" ["4" "5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A6" ["3" "4" "5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A7" ["3" "4" "5" "6"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A8" ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "A9" ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "22" ["4" "5" "6" "7"] "P")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "33" ["4" "5" "6" "7"] "P")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "55" ["2" "3" "4" "5" "6" "7" "8" "9"] "DD")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "66" ["3" "4" "5" "6"] "P")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "77" ["2" "3" "4" "5" "6" "7"] "P")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "88" ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"] "P")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "99" ["7" "10" "A"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "1010" ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"] "S")
(update-cheat-sheet-cell row-names col-names blackjack-cheat-sheet "AA" ["2" "3" "4" "5" "6" "7" "8" "9" "10" "A"] "P")
